scripts/coverage_report_jenkins.py

```diff
@@ -374,106 +386,215 @@ def sonar_scan(ssh, project_name, project_version, scanner_url, scanner_dir, ser
             continue
 
+        # We have all the data to instantiate our Build object, so lets do it.
+        # Note, we could fill out its data members a little at a time because
+        # a namedtuple's data members are immutable.
+        build = Build(
+            number=build_number,
+            job=jenkins_job,
+            coverage_archive=artifacts['coverage-results.tgz']['relativePath'])
+
         if not check_artifact(
-                jenkins_user, jenkins_token, jenkins_url, job_name, build_number,
+                jenkins_data.user,
+                jenkins_data.token,
+                jenkins_data.url,
+                jenkins_job,
+                build_number,
                 artifacts['coverage-results.tgz']['relativePath']):
-            logger.info('Coverage archive not possible to be downloaded, skipping')
+            logger.info('Coverage archive could not possibly be downloaded, skipping')
             continue
 
-        if build_appliance_version == appliance_version:
+        if build_appliance_version == cfme_version:
             logger.info('Build %s was found to contain what is needed', build_number)
-            eligible_build_numbers.add(build_number)
+            eligible_builds.add(build)
         else:
             logger.info(
                 'Skipping build %s because it does not have correct version (%s)',
                 build_number,
                 build_appliance_version)
 
-    if not eligible_build_numbers:
-        raise Exception(
-            'Could not find any coverage reports for {} in {}'.format(appliance_version, job_name))
+    return eligible_builds
 
-    eligible_build_numbers = sorted(eligible_build_numbers)
 
+def setup_appliance_for_merger(appliance, ssh):
+    """Setup appliance for code coverage merger
+
+    Stops the evm service, and then installs the coverage_merger.rb
+    script and dependent libraries.
+
+    Args:
+        appliance: CFME appliance object
+        ssh:  ssh object
+
+    Returns:
+        Nothing
+    """
     # Stop the evm service, not needed at all
     logger.info('Stopping evmserverd')
     appliance.evmserverd.stop()
+
     # Install the coverage tools on the appliance
     logger.info('Installing simplecov')
     appliance.coverage._install_simplecov()
+
     # Upload the merger
     logger.info('Installing coverage merger')
     appliance.coverage._upload_coverage_merger()
-    with appliance.ssh_client as ssh:
-        if not ssh.run_command('mkdir -p {}'.format(coverage_dir)):
-            raise Exception(
-                'Could not create coverage directory on the appliance: {}'.format(coverage_dir))
 
-        # Download and extract all the coverage data
-        for build_number in eligible_build_numbers:
-            logger.info('Downloading the coverage data from build %s', build_number)
+    if not ssh.run_command('mkdir -p {}'.format(coverage_dir)):
```

Comment:
General suggestion for your `ssh.run_command` calls where you raise exception - use an assert to make the statement more concise, and to raise a specific exception.

```python
assert ssh.run_command('rm -rf /*').success, 'AssertionError context message goes here'
```

Reply:
We ended up in a long internal discussion.   The end result is I have a wrapper function for ssh.run_command() that handles logging command ran, exit code, and output of the command, and throwing an exception with a specified error message.